File: src/main/java/com/timeback/ui/domain/gateway/FeatureGateway.java
package com.timeback.ui.domain.gateway;

import com.timeback.ui.domain.model.*;

import java.util.concurrent.CompletableFuture;

/**
 * CT-04 기능 호출 계약 — frontend-components.md §4~§11 매핑
 * Java에서는 CompletableFuture로 비동기 호출을 표현한다.
 */
public interface FeatureGateway {

    // --- UI-01: 권한·초기 진입 ---
    CompletableFuture<ScreenState> readAccessState();
    CompletableFuture<ActionResult> openUsageAccessSettings();
    CompletableFuture<ScreenState> refreshAccessState();

    // --- UI-02: 홈 대시보드 ---
    CompletableFuture<ScreenState> readHome();
    CompletableFuture<ScreenState> refreshHome();

    // --- UI-03: Timeline ---
    CompletableFuture<ScreenState> readTimeline(long date);
    CompletableFuture<ScreenState> refreshTimeline(long date);
    CompletableFuture<ActionResult> createActivity(ActivityType type, String name, TimeRange timeRange);
    CompletableFuture<ActionResult> updateActivity(Identifier activityId, String name, TimeRange timeRange);
    CompletableFuture<ActionResult> confirmContext(Identifier contextId);
    CompletableFuture<ActionResult> updateContext(Identifier contextId, ContextClassification classification);

    // --- UI-04: 앱 관리 ---
    CompletableFuture<ScreenState> readApps();
    CompletableFuture<ActionResult> changeDefaultClassification(String packageName, AppClassification classification);

    // --- UI-05: 시간 되찾기 ---
    CompletableFuture<ScreenState> readRecoveryEntry();
    CompletableFuture<ActionResult> startGoalTimer(Identifier goalId);
    CompletableFuture<ActionResult> completeGoalTimer();
    CompletableFuture<ActionResult> createManualRecoveredTime(Identifier goalId, TimeRange timeRange);
    CompletableFuture<ActionResult> selectRepresentativeGoal(Identifier overlapGroupId, Identifier goalId);

    // --- UI-06: 목표 ---
    CompletableFuture<ScreenState> readGoals();
    CompletableFuture<ActionResult> createGoal(String name, Duration targetDuration);

    // --- UI-07: 리포트 ---
    CompletableFuture<ScreenState> readReport(ReportPeriod period);

    // --- UI-08: 데이터 관리 ---
    CompletableFuture<ScreenState> readDataManagementState();
    CompletableFuture<ActionResult> changeRetentionSelection(RetentionSelection selection);
    CompletableFuture<ActionResult> requestFullDeletion();
    CompletableFuture<ActionResult> confirmFullDeletion();
    CompletableFuture<ScreenState> refreshDeletionStatus(Identifier jobId);
}

File: src/main/java/com/timeback/ui/domain/gateway/FeatureGateway.kt
package com.timeback.ui.domain.gateway

import com.timeback.ui.domain.model.*

/**
 * CT-04 기능 호출 계약 — frontend-components.md §4~§11의 조회·작업 매핑
 * track-ui는 이 인터페이스만 사용하고 실제 구현을 알지 못한다.
 * FakeFeatureGateway와 실제 FeatureGateway 모두 이 인터페이스를 구현한다.
 */
interface FeatureGateway {

    // --- UI-01: 권한·초기 진입 ---
    suspend fun readAccessState(): ScreenState
    suspend fun openUsageAccessSettings(): ActionResult
    suspend fun refreshAccessState(): ScreenState

    // --- UI-02: 홈 대시보드 ---
    suspend fun readHome(): ScreenState
    suspend fun refreshHome(): ScreenState

    // --- UI-03: Timeline ---
    suspend fun readTimeline(date: Long): ScreenState // date = epoch millis of local midnight
    suspend fun refreshTimeline(date: Long): ScreenState
    suspend fun createActivity(type: ActivityType, name: String, timeRange: TimeRange): ActionResult
    suspend fun updateActivity(activityId: Identifier, name: String?, timeRange: TimeRange?): ActionResult
    suspend fun confirmContext(contextId: Identifier): ActionResult
    suspend fun updateContext(contextId: Identifier, classification: ContextClassification): ActionResult

    // --- UI-04: 앱 관리 ---
    suspend fun readApps(): ScreenState
    suspend fun changeDefaultClassification(packageName: String, classification: AppClassification): ActionResult

    // --- UI-05: 시간 되찾기 ---
    suspend fun readRecoveryEntry(): ScreenState
    suspend fun startGoalTimer(goalId: Identifier): ActionResult
    suspend fun completeGoalTimer(): ActionResult
    suspend fun createManualRecoveredTime(goalId: Identifier, timeRange: TimeRange): ActionResult
    suspend fun selectRepresentativeGoal(overlapGroupId: Identifier, goalId: Identifier): ActionResult

    // --- UI-06: 목표 ---
    suspend fun readGoals(): ScreenState
    suspend fun createGoal(name: String, targetDuration: Duration): ActionResult

    // --- UI-07: 리포트 ---
    suspend fun readReport(period: ReportPeriod): ScreenState

    // --- UI-08: 데이터 관리 ---
    suspend fun readDataManagementState(): ScreenState
    suspend fun changeRetentionSelection(selection: RetentionSelection): ActionResult
    suspend fun requestFullDeletion(): ActionResult
    suspend fun confirmFullDeletion(): ActionResult
    suspend fun refreshDeletionStatus(jobId: Identifier): ScreenState
}
